test/mpi/comm/cmsplit_type.c

```diff
@@ -55,6 +55,67 @@ int main(int argc, char *argv[])
         MPI_Comm_free(&comm);
     }
 
+    /* Check to see if MPI_COMM_TYPE_HW_GUIDED works correctly */
+    for (i = 0; split_topo[i]; i++) {
+        MPI_Info_create(&info);
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[i]);
+        MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+        if (comm != MPI_COMM_NULL) {
+            int newsize;
+            MPI_Comm_size(comm, &newsize);
+            if (newsize > size) {
+                printf("Expected comm size to be at most the node size\n");
+                errs++;
+            }
+            MPI_Comm_free(&comm);
+        }
+        MPI_Info_free(&info);
+    }
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: pass different info values, it must return
+     * MPI_COMM_NULL.*/
+    MPI_Info_create(&info);
+    if (rank % 2 == 0) {
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[0]);
+    } else {
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[1]);
+    }
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("Expected MPI_COMM_NULL due to different info values.\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    MPI_Info_free(&info);
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: pass info to some processes, it must return
+     * MPI_COMM_NULL.*/
+    if (rank % 2 == 0) {
+        MPI_Info_create(&info);
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[0]);
+    } else
+        info = MPI_INFO_NULL;
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("Expected MPI_COMM_NULL due to inconsistent infohints.\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    if (rank % 2 == 0)
+        MPI_Info_free(&info);
+
+
+    /* Check to see if MPI_COMM_TYPE_HW_UNGUIDED works correctly */
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_UNGUIDED, 0, MPI_INFO_NULL, &comm);
+    if (comm != MPI_COMM_NULL) {
+        int newsize;
+        MPI_Comm_size(comm, &newsize);
+        if (newsize >= world_size) {
+            printf("Expected comm to be a proper sub communicator\n");
+            errs++;
+        }
+        MPI_Comm_free(&comm);
+    }
 
```

Comment:
Can we also add a test to check the returned info with `MPI_COMM_TYPE_HW_UNGUIDED`?

>If a valid info handle is provided as an argument, the MPI implementation sets the info key "mpi_hw_resource_t ype" for each MPI process in the group associated with returned newcomm communicator and the info key value is an implementation-deﬁned string that indicates the hardware resource type represented by newcomm

Reply:
done.